### ISSUE TITLE
Poor conditional type performance with large string literal union types

### ISSUE THREAD

**Issue by phulin** [microsoft/TypeScript#47481]
# Bug Report

The TypeScript compiler has poor performance when compiling conditional types involving large string literal unions. In particular, performance appears to be multilinear in the size of the unions involved. The example below takes 19 seconds to compile on my (recent) laptop, with two 10,000 string literal union types. A 5,000 x 5,000 example takes 4.5 seconds, a 10,000 x 5,000 example 9 seconds, and a 15,000 x 10,000 example 27 seconds, suggesting the asymptotic performance I indicated.

I discovered this issue while attempting to implement strong typing on requests to a key-value store in a project. The values in the store have different types that are static, but the store is large enough that using types like the example below leads to what is, for me, unacceptably slow compile times (which are then reflected in slow VS Code Intellisense updates).

I imagine this may be a fundamental algorithmic limitation, but I don't know enough about TS's inner workings to know.

This may be related to #29350.

Thanks to all TS contributors - hope this report helps.

### 🔎 Search Terms

union conditional type poor performance slow multilinear quadratic

### 🕗 Version & Regression Information

Observed on multiple versions, including today's `next`. I have not seen a version with good performance in this case.

### ⏯ Playground Link

Unfortunately, playground screeches to a halt with the full 10,000 x 10,000 test case. Here is a 1,000 x 1,000 example instead:

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.6.0-dev.20220116#code/KYDwDg9gTgLgBDAnmYcByBXAtsKBLAYwAUoIVZE4BeOAIgDsBGWuAHzvoCYX2GBmHhwAsghgFZR9AGySA7JIAckgJyTGABjXM2HRtx0NGAg0xEnGE8zPPzzS86pOdNT7by773nY17NfLXtZetl72Xo7ufC6Rbhx8nnE+cX5xAXFBcSFxYXERwtHCsQxCCcVJxSnFacUZxVnFOcV54gXiRfRipR3lHZUd1R21HfUdjR3N0q3S7VJdUj1SfVIDUkNSI1JjUhOyU7Ltsl2yPbJ9sgOyQ7IjsmOyEwpTCu0KXQo9Cn0KAwpDCiMKMYKCbKKbKdrKLrKHrKPrKAbKIbKEbKMbKCYaKYadoaLoaHoaPoaAYaIYaEYaMYaDEaLQ4vRaAmMIkWLRkxgUxhUxgY5xqTg4zh47z8omcEmcMmcCmcKmcDFRNR8HHxJUEvhEvgkvhkvgUvhUvgYoRYoQ4kpqIQEoREoQkoRkoQUoRUoQYsRYsQ4zpqMQEsREsQksRksQUsRUsQYqRYqQ42ZqeaJonLRNk9aJqnbNS7HM4w45gmnHMky45im3HMYx5qZ61vHvWtE761sn-WtU4FqUHdnGQ7sE2HdkmI7sU1HdibOKbOdrOLrOHrOPrOAbOIbOEbOMbOKe01xzhmuJfMyQCteMDccs9cm9TvlOAVnoXPpecFfi58b6XPnfys+Kk4yoAQufBLhqAFrjqAFbgaAFTiaZ5mkhC5WkhK52khG5OkhO5umeHoEXOPpOH6BErkGBEbmGBE7lGZ4xgxc4Jk4SasSuqasRuGasTu2ZOLmAlzgWAlLsWAlrmWAlbpWAlTjWTh1opC6NopK4topG7topO5dk4Pb6XO-b6UuQ76Wuo76VuE76RMURTFE7RRF0UQ9FEfRRAMURDFEIxRGMUR2fuMROUeMRuaeJjKl5l6SMqfm3lFPJxQ+kRPlFL4ZW574pV5kopX5sopXZgGRMBUWqhVbkQRVXnQRVflwRVdmIVFyFtS5aFtR5mFtT5OFtQF+FRYRI1OSRkRkSNHmUSNPk0SNAX0VFjErU5LGRGxm0eZxm0+Txm0BfxkSCSdTkiSdbniSdXlSSdfmySddkKZESmvS5qmvR5GmvT52mvQFemRAZwNOcZwNuWZwNeZZwN+TZwMTCaUwmu0JpdCaPQmn0JoDCaQwmiMJpjCaSPBYUaNhYUWORe4Zp47FJhmkTiV08lTOpcI6V05lPNYzlHN4-lHNE0VHNI6VwjlXTlUy1jNUy3j9Uy0TTUy0jrV0+1WsY11Ws471WsEwNWsk8NdOjRbaMTcIU0Wzjs0WwTC0WyTy106tHtoxtwhbb7OO7b7BMHb7JPHcIp0R2jF0R1j10R3jd0R0Tj0R0jL3CG9mcY59mc4z9mcE-9mck0Dwgg+XaPg+XWNQ+XeOw+XRMI+XEwelMHrtB6XQej0Hp9B6AwekMHojB6Ywem35NtF3VNtH3tMcF6Q+M+4Xpj6zS-s2vnPiNzS+8wffcCzvQ-CzvY9izvbeS+I0tL7LD99wrD9D8rD9j2rD9t5rS-a3-Pc9Z-wHobP+I8TZ-wnubJelsYFdxtuIO2MCB6OxgSPF2MCJ7uyXp7HBXcfbiD9oQgegdCEjxDoQie4dxCRxoV3GONC+7xxoUPJONCx6pxoW3DO4gs68J7rnXhA8C68JHsXXhE8y7iArtIru1dpF9zrtIoejdpFjxbtIiYMYpgxhmOoOY6gFjqCWOoFY6g1jqA2OoLY6gtHT2mDMOe0wFiLwYHGFYq8OBxg2JvNx28vG72kPvNxh8QkLBPgElY58AkbCvgErRt9pD3zcY-FJCwX4pJWO-FJGwv4pK0b-Nx-8ilzCAUUpYoCilrAgUUrY0C3GwIaTMBB0gkENKWKghpawMENK2NgtxuCBkzAIdIIhoylikNGWsChoytjUOkLQhZMwGELIWMwhZKw2ELI2JwhZWieHSD4YcuYgjDlLBEYctY4jDlbCkdIGR9yZjyPuQsJR9yViqPuRsDR9ydjqD2OoA4+i5CGJBWcUxIKriWJBXcWxch7H7AOE4-YJxXH0H2BcTxDB9g3F8ei-x2LAmHAOKE9FIoTCHDOJ+Cl0TCU3DiYSnYiTjgHFSeisCcgMnsouNk9lNw8nsp2IU9FxSRVHDKSKs4lSRVXBqSKu49T0WNKVQcFp5wTgBjkJ0pVVwelKruP09FgyjUHBGZcE4iw5CTMuFcGZlw7jzOuHsfYcgVnXBOOs64FwtnXBuLs64OwDm3AOK8OQpzbhnAubcK41zbh3DufcPY4I5DPPuCcN59wLifPuDcH59wHj-MUICotbxQUmEeF8CF5bzFFoBNYotDx7HPBeE454Hw0XPB+Fi+gzwAR4ueA8QJrwXikteB8CJDBXg-FpT2n85aGWzoeIk94Lw2XvA+Fy94PxeXvABAK94DxhWfBeBactErPhfGlZ8P4crPhAkVd8J4XpFAtO+B8TV5btXfD+Hq74QJDW-CeHGRQIzfgfEteWyZvw-gzN+ECeZ-wnguvLSs-4Hx1n-B+Fs-4AJdn-AeAcwELxQ3ltOYCL4FzAR-GuYCIEdzgRPGTeW55wIPhvOBD8T5wIAQ-OBCCQtJhQQQmBYJst7hQRwireJmtomUT1tEyCex4IIROPBDCNF4IETdvBCiPF4IQSBMhBCUlkIYQTvoJCBEM7IQogXZCEEiToQQjZdCGEXLoQIl5dCFEAroQgmFbCCEp7xMSthHCaVsIkRythGiRV8IwTPsEy0+EMIP3ie1fCJEer4RokNYiMEwHBMjMRDCCD4nJmIiRDMxEaJ5nIjBMh8TKzkQwnWciBEWzkQol2ciEEBzUQQhI+J05qI4QXNREia5qI0R3PRGCJj4nnnohhG89ECJPnohRD89EGJ-maAANwAChQCQFgAgZAqAADKMB8D0AAOYkDILgJA1A6AAGcihvdKG98ob3KhveqG92ob36hvcaG95oH3Wgfc+3PD7v3XEfcB1ij7oPfEfch7vL7n3Qlfd+xOr7gPaVfdBwyr7kPb4-c+6kn7v2Mk-cB9kn7oO8k-ch4Uv7n3gvvbKX9-7lS-vA5qX98H9SAfQ8S7wAH322kA-+50gHwOekA-B-0oH0PCtS4IUD37ZX3ukKB8DihQPwfUJB9Dxr72GEg9+8wkHgO2Eg9B5wkHkOeFg8+0N97giwf-ZEWD4H4iwfg6kRD6HC33vyIh79pREPAeqIh6DjREPIeYlEFD2HImpf4nT4SXPUn3uknz2j+T2e4UGBh9Dxg1fc9w+rwjxH1fkco+r2j9H1fU9Y70LD3HegEcE70Mj4neg0dk70KnynRhYc06MAj+nRhkdM6MGj1nRhU8c+ZLD7ntAPu8+ZP95kyOHS5+F8ycHzJU+NKR7DhBSOEdpcLwriwwOLBo4jLntXl4q+a8L9ry8CO9cd9LxkdVhc9jdLxz8zcOQq9LcgDrcOQEc7cORkdHcOQ0cXcORU93cuRYcvcgCfcuQD9-cuQX8g8uRz9Q8eQq8I8gCo8eQEdY8eRkcE8eQ0dk8eRMcBMpdZx095xeCxN3tlx+CicZNuDoUK9tx+DMdp5sccc4cBR8dEdzxeCUcBRSd0cBRMcschQcdcchR8cCchQidichRScychRMdKdvAccadvB8d6dvAicmdvBSdWdvBMcOd3wcdt8vted3x-t3widj8JDhd3xwd3xMcr9xQcdb9xR8cH8d9xQicQxeDldxRwi1dJRodJQcdtdJR8dACvsDdJRgdJRSdNheCzdpRsjYCSdvtpR8c7dpQidHdpRScXdpRMd3dZQcc8CvsfdZQAj-dZRSig9ZRwjQ95RsiaDyd6joReDY95QicE95RSdk95QKcuD3tHJ09nJdiBCd93J9jGdRDtjxCpd-J9iKcZDlRqc4dlQ6dEdopdiUd4oXjwdlQKcsd4hqdcd4g6cCd4hGdid4gWcyd4gKdKcwJqcacwI6d6cwJGcmcwIWdWcwIKcOcNRqcfCNQ6cbRdiBcNRgcNQWcXQCSKcr8tRqdb8tQ6cEiGdGcUiK8tQWd39mS1cdRocdRqdtcdQ6dCidRGdQDmTjcdQPizc9QuTai9Rvs9Q6c7c9RGdHc9QWcXc9QKd3cDRqc+iDQ6dPhdj-cDRiSg8DQPjQ8jQuSZijRZT5jmTY8jRGcE8jQWdk8jR2ctid9UZ090ZvSDi-tjFfTAd8ZfTQdiZfT2cZCzQuc4czRfszR+dWQK8zQhdrwkz0czR2cscSgudccSg4yCcSggzicShQyycSh2dKcrQucacrQ4z6crQgymcrRQzWcrR2cOcbQucfCbQ4z8SkyBcbQhdhcbRRcxc7Roc7Qudb87Q4z6S7QgymSpc7RQy2Sly1cHQJyf9PTtcHQ4zCiHQgzhSlzjcHRRczcnQJzainRvsnQ4y7cnQgzHcnRQyXcnR2d3cXQuc+iXQ4z9Skz-cXQhcg8XRRdQ83QJyZi3QbzbSlzY83QgyE83RQzk83RId2509O4MKs93te4sL5cC8d9h4sLQdx4sK0KZCvRPsvQZcjAMLEdl46KldUypcvRVcCVpcJdglpcZdyUWKCdOhAdOglc50+LVd-wK8-QJdkkAcac-Rfs-R5ctQMKmc-QSLWc-Q0KOcAwqKfCAx5K+yWKBcAwldhcAxVcxcgwJdJccLb8gx5L6SgxBLFybKSLVybK0LBlFcqLtcQx5LCiQxBKjycLjcQxVczcwwJdaiwwZdjgMK7cwxBLHcwwSKXcww0L3cIwqK+iIx5K-yWL-cIwlcg8IxVdQ8owJcZiowZcYKcLY8oxBKE8owSLk8oxIdtF09dEOrsKd8YxdcAyK8YxAcYwjdzj9dS8xq2qZC4xPs4xvs4xddEd3EOqUdvFlrTd2LZgNcuLZg5reL9cCdZghridZhQdZhTdxKtdKd5gZqad5hdd6d5ghqmd5hTrWd5g2qOdFgZqfDFhdcDL9cBdFgjdhdFhTcxdlgNdrKerb9lhdd6TlghrnLobTq3Loa2rPLVgZrtdVhddCjVghqgqerjdVhTczd1gNdaj1g5rYqBq7d1ghrHd1hTqXd1g2r3dNgZq+jNhdc8r9d-dNgjcg9NhTdQ9tgNcZjtg5qaqerY9tghqE9thTrk9thIddgLdi0K9dhvtdhbd+qpddgHdTid9dhndxrjby99aZDEV099htbaLNbEcMUbaUccVnbwd9hVasdiUbbcdDhbcCdDgHdidDhncydDhVbKcWUbaadjhbd6djgHcmdjhndWdjhVaOdThPtThtbedTh-tTgHdgj9bhdTh3axdzgLcob7dtbZdzh86FdzhgdzhndUb7dVbPLLgs7tdzUbbCjLgHdCaQdjd7Ve7VbaEncs7rd3Uba7cvUZ6m7rgZ73b7gbb3dg1V7tafcI1V6HdfhV7ndARV7VaZFXcs6o800bbY9M1L6m7kRL73b0R09HhodHhPduqwc-SK0n6CKP7A9Rqd9HgQ8LbvcZDm0n64c21wG-dEypdnhA9mKQGQ92LXgX6uLXhvsx0n6Ccp0sHA8RLvcydXg3dKcV0n6ad10yG-clKK93hA89QyGQ8jQn6Odj1mGMHecL1mHAdr1mHQc71mG3cr9vhPdb830n76TvhuGkawdlc-1xG3dPLfhPdtcwMn7CjfhuHB7fg+GKiaGzdEMn7aj-gMGabYG7dMNDHA9F6aGXd8Mn73ciN7GMGfdyN7HuG96aGg9aN7G3cT7gRPco9WMn7Y8ONgnA876aHk8+N09QRw8NapdQRvtQQY89bI8f7QRgdQQk8zaw9IdwRw8a8K9wQkn7aEnEdNMYmUcdNKnwd9MYmscjN6mkm9qd9IR-tLN6nMn8HWmyd7MYnKcnN+mkmOUin6cPN+nMn6HRnammGimOdAsYmfDYQY9-rWmBdItFmk8yS5m8mr94RPt4QknZd4R2mFdMsYnlccsLm8nPLEQDntcSsYnCjER49B7EQk9dGEmzd6sYnajkQknTHI87d2tfnMnrGEmXdesYn3cBtoWkmfdRtoX48PGEmg9ptoW8mT70QDmo9lsYnY81t8XMmImEnk8dt89C0jtjtwBoB4AkAUA4AABpegCAAAd3oEe3IBeyoEOzgB0EwBwHwGIFIC5cQF5Z0Gu1uwexFeezFcO3pdQE5dlYADUABDAAGwwGAAAB5xWlWKAAAaQ7AAPle31Ze1ABgGAHoAABM3t0BsBcBCBzWxW+WAB+OAegbAAAI1wHFYAC44AXW4BLXrW7W4BJW8B7sXXxWPW3sbso27s4BA2vX1X1WqWTtaW4AAAzDAegAgGAPACAegOADAN7YAF17V4N0N21+15ltljlmVigY1gACnFZteAGzdVYwHVZgAAH1A2XW1XNWdWXXjWjWABKQdptpAYdrVqtmdxAU1gAb3FagGABgAwCgBLY7a7Z7f7aOwAF9DsgA)

[Test case repository](https://github.com/phulin/repro)

### 💻 Code

```ts
export type NumericProperty = "n1" | "n2" | /* snip */ "n10000";
export type StringProperty = "s1" | "s2" | /* snip */ "s10000";
export type KnownProperty =
  | NumericProperty
  | StringProperty

type PropertyValue<
  Property,
> = Property extends NumericProperty
  ? number
  : Property extends StringProperty
  ? string : null;

export function useProperty<Property extends KnownProperty>(
  default_: PropertyValue<Property>,
): PropertyValue<Property> {
  return default_;
}
```

**Comment by tinganho** [comment]
I work on a project where we use a lot of translation strings. We have provided a union type of all translation keys. Some times we have variables that just works on a sub category of the union strings and we use `Extract<T, U>` to extract them. We have noticed that the project just explodes in compile time. From 36s to over 4 minutes. TS Server is also very sluggish.

We have used `--generateTrace` flag to zero in on the expressions that takes long time. And we concluded it to be `Extract<T, U>` that is the culprit.

I created a repo of the perf bug we experienced. If you want to evaluate.
https://github.com/tinganho/extract-bug

We found the following quote in https://github.com/emotion-js/emotion/issues/2257
From @amcasey:
> Why don't we topologically sort them? I wasn't there but I would guess it was either too hard in the general case or didn't seem like it would make a difference (this is a pathological case).

Maybe our case is very specific. But, I think it would make sense to sort them based on a certain threshold? At least, I think it's quite general that the more union constituents you have the more lookups people tends to do.
